### PR TITLE
7856 - Shell BackButtonBehaviour TextOverride breaks back navigation

### LINKED ISSUE #7856: [Bug] Shell BackButtonBehaviour TextOverride breaks back navigation
### Steps to Reproduce

1. Override text in a BackButtonBehaviour
ex. Shell.SetBackButtonBehavior(this, new BackButtonBehavior
            {
                TextOverride = "Test"
            });

### Expected Behavior
The Back button should now have the text "< Test"

### Actual Behavior
The Back button has the value "Text" (without the arrow) and opens the Flyout menu rather than navigate back

### Basic Information

- Version with issue: 4.2.0.778463
- Last known good version: Unknown
- IDE: Visual Studio Enterprise 2019
- Platform Target Frameworks: 
  - iOS:  Xamarin.iOS 13.2.0.42

### PR DESCRIPTION
**Pull request by dhindrik**
### Issues Resolved ### 
- fixes #7856

### API Changes ###
 None

### Platforms Affected ### 
- iOS

### Behavioral/Visual Changes ###
It now shows back button behaviours as it should.

### Testing Procedure ###
Create a Shell and with one page and set BackButton Override. Navigate to another page and check that the back button is shown as you overrode it.

Video:
[7856.mp4.zip](https://github.com/xamarin/Xamarin.Forms/files/3978657/7856.mp4.zip)

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by dhindrik** [comment]
I will rebase and add a test. I am on holiday this week, but I will do it at latest in the middle of next week.

**Comment by dhindrik** [comment]
Looks like I did something wrong with the rebase?

**Comment by PureWeen** [comment]
if you do this

- git checkout 4.5.0
- git pull
- git checkout 7856-back-button-override-ios
- get rebase 4.5.0
- now force push then does that fix?

If you don't have 4.5.0 in your local repository then you will need to sync it with our upstream 4.5.0 branch
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

**Comment by dhindrik** [comment]
> if you do this
> 
> * git checkout 4.5.0
> * git pull
> * git checkout 7856-back-button-override-ios
> * get rebase 4.5.0
> * now force push then does that fix?
> 
> If you don't have 4.5.0 in your local repository then you will need to sync it with our upstream 4.5.0 branch
> https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

Thank you! Looks like it worked. Even if my commits are there multiple times.

**Comment by PureWeen** [comment]
oop sorry @dhindrik it looks like you didn't complete the rebase

You have to fix the conflicts and then do a 
git rebase --continue

The VS git tools are pretty good for this or you can use
https://git-fork.com/

**Comment by dhindrik** [comment]
I did complete it, but the things I did the first time, probably messed things up. But now I did an interactive rebase and dropped all my commits except the "real" ones.

**Comment by dhindrik** [comment]
@PureWeen are there more changes that I need to do?

**Comment by dhindrik** [comment]
Looks like there are some problems with the UI-Tests and the Gates of the stages.

Request body: {}
				Response Code: 0
				Response: An error was encountered while processing request. Exception: The specified CGI application encountered an error and the server terminated the process.
Exception Message: The remote server returned an error: (502) Bad Gateway. (type WebException)
				Evaluation of expression 'in(root['exit_code'], '0', '1')' failed.
				
Exception Message: Error parsing API response into a JSON object. (type DistributedTaskException)

**Comment by dhindrik** [comment]
@davidortinau @PureWeen @samhouts Will this be part of 4.5?

**Comment by dhindrik** [comment]
Rebased with 4.5.0!

**Comment by dhindrik** [comment]
Rebased! :)

File: .nuspec/Xamarin.Forms.DualScreen.nuspec [hunk 1]
```diff
@@ -1,56 +0,0 @@
-<?xml version="1.0"?>
```
Comment (PureWeen) [inline .nuspec/Xamarin.Forms.DualScreen.nuspec]: Fix me being deleted

File: Xamarin.Forms.ControlGallery.Android/Xamarin.Forms.ControlGallery.Android.csproj [hunk 2]
```diff
@@ -55,7 +55,6 @@
     <WarningLevel>4</WarningLevel>
     <AndroidLinkMode>None</AndroidLinkMode>
     <EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>
-    <AndroidUseSharedRuntime>false</AndroidUseSharedRuntime>
```
Comment (PureWeen) [inline Xamarin.Forms.ControlGallery.Android/Xamarin.Forms.ControlGallery.Android.csproj]: Don't remove

Comment (dhindrik) [reply Xamarin.Forms.ControlGallery.Android/Xamarin.Forms.ControlGallery.Android.csproj]: I have added it again now

File: .nuspec/Xamarin.Forms.DualScreen.nuspec (hunk 1 above)
Comment (dhindrik) [reply .nuspec/Xamarin.Forms.DualScreen.nuspec]: it is back :)

**Comment by dhindrik** [comment]
I think I have resolved those issues now related to my earlier commit issues.

**Comment by dhindrik** [comment]
I have fixed so that UI-test is working! I have 3 other UI-test failed when running them locally. But for me, it doesn't seem related to my change at all. 

It is Issue488TestsLongTextRotation, Issue5470Test and VisitAndCheckForItems (CollectionView).

**Comment by PureWeen** [comment]
@dhindrik I wouldn't worry about those 

Kicked off UI Tests 🤞

**Comment by dhindrik** [comment]
Looks like it failed when trying to create the NuGet package. Is that related to my changes?  I can't see what I changed that could make that step fail.

**Comment by PureWeen** [comment]
@dhindrik can you rebase? that should fix

**Comment by dhindrik** [comment]
Yes! It is rebased with 4.5.0 now.

**Comment by PureWeen** [comment]
https://github.com/xamarin/Xamarin.Forms/pull/8951#issuecomment-594173779

**Comment by PureWeen** [comment]
Go UI Tests Go!!!

**Comment by PureWeen** [comment]
API19-FR unrelated

**Comment by PureWeen** [comment]
@dhindrik it looks like this breaks some of the behavior now when you've pushed a page

Control gallery => Go To Test cases => Shell Back Button Behavior Test => Push Page

- Toggle Text doesn't work
- Toggle Is Enabled doesn't work

as you fix these do you have time to add UI tests for these scenarios as they are obviously missing? If not let me know and I can add them

**Comment by dhindrik** [comment]
The reason that toggle text not working after you pushed a page is that I am using the BackButton of the navigation item. This to get the iOS default behavior that you set a back title for a page, and that title  will show like the back button when you are navigating to next page.

**Comment by PureWeen** [comment]
@dhindrik Yea I think being able to set that text is pretty important. 

I think having BackButtonBehavior do anything with the flyout was a mistake and if I could go back in time I'd have created a FlyoutButtonBehavior that just let you configure it separately from the backbutton. 

But one of the main things the BackButtonBehavior should do is let the user set the Text on that button when there's a page pushed

**Comment by dhindrik** [comment]
But how will I be able to set the text and keep the default icon? Maybe add a new property for that? Or a platform-specific just for iOS? I will be able to use just a back-arrow, the same behavior as with NavigationPage.SetBackButtonTitle (this, string.Empty); But that maybe is should create a new issue and a new PR for that and fix this to work as you say and how it works on Android?

**Comment by dhindrik** [comment]
How to take this further?

**Comment by PureWeen** [comment]
@dhindrik Tell me what you think of the latest changes I pushed. It still needs a little refinement and testing.

The main thing (that I always forget) is that the BackButtonItem used is the one on the previous controller. So the text wasn't working because it was being set on the wrong BackButtonItem instance. You have to set the `BackButtonItem` on the previous ViewController

**Comment by dhindrik** [comment]
> The main thing (that I always forget) is that the BackButtonItem used is the one on the previous controller. So the text wasn't working because it was being set on the wrong BackButtonItem instance. You have to set the `BackButtonItem` on the previous ViewController

Not if you want it to work like NavigationPage.SetBackButtonTitle, but it is more obvious how to work with it if you can change it on the page that you are at. At least for new developers that haven't used the NavigationPage.SetBackButtonTitle method before.

**Comment by PureWeen** [comment]
@dhindrik that's how it currently works so that's how it has to continue to work

We would have to break the current behavior of BackButtonBehavior on Android as well.

Either way, I feel like it'd be awkward to have the only API on BackButtonBehavior that applies to the next page is text. 

The Icon and the command apply to the current page but the text is the only property that applies when the page is pushed? That feels really confusing.

If we went that direction then we'd make it so the entire BackButtonBehavior applied to A would be what you used when you pushed to B

But the ship has sailed to do it like that :-)

**Comment by PureWeen** [comment]
> I think the right way is to use BackBarButtonItem to set the back button text and LeftBarButtonItem to customize the menu icon. But if you want I can spend some time to find a solution to it.

I totally agree with this but that'll need to be a different PR

for now we need to just focus this one down to fixing the TextOverride weirdness.

I think we'll eventually we will break all this stuff up 

<HamburgerToolbarItem
<BackButtonToolBarIItem
<LeftToolbarItem

**Comment by PureWeen** [comment]
All the UI tests are green on this one. Not sure why github is reporting api27 as X

**Comment by dhindrik** [comment]
> 
> Either way, I feel like it'd be awkward to have the only API on BackButtonBehavior that applies to the next page is text.

Yes it is. I just expected that behavior because of how it works with NavigationPage.

**Comment by PureWeen** [comment]
@dhindrik Yea I always forget that :-/ obviously

Which also makes me realize I was completely missing most of the points you were trying to make here. 👎  

Unfortunately this won't be part of the next SR for 4.6 but should be part of the next next one.

Thanks for all your work and collaborations as always!!

**Pull request merged**